src/Cardano/Wallet.hs

```diff
@@ -301,22 +305,23 @@ walletNameMinLength = 1
 walletNameMaxLength :: Int
 walletNameMaxLength = 255
 
-newtype WalletId = WalletId Text
-    deriving (Eq, Ord, Show)
+newtype WalletId = WalletId UUID
+    deriving (Generic, Eq, Ord, Show)
 
-newtype WalletTimestamp = WalletTimestamp Microsecond
-    deriving (Eq, Ord, Show)
-
-data WalletStatus
+data WalletState
     = Ready
-    | Restoring
-    deriving (Eq, Show)
+    | Restoring !(MeasuredIn "percent" Word)
```

Comment:
@Anviking @jonathanknowles 

The `Percentage` here is giving me some hard-time and there are a few points I'd like to clarify:

- The `MeasuredIn` (or `Quantity`, this is probably indeed a better name) is a polymorphic type which allows us to represent numeric quantities with units. It's not _only_ about the wiring to JSON. It happens that we do want consistency here and represent all those numeric types in the same way. It gives extra meaning to a type. Like `Key` or `Index` from the Address derivation module.

- There's nothing polymorphic about `MeasuredIn "percent" Word`. This type has only one and one concrete representation and is strictly equivalent to a `newtype Percentage = Percentage Word`. I actually went for the latter initially but it felt _wrong_ since, we already had a type to represent quantities but, we were wrapping things in a `MeasuredIn` during the serialization and deserialization. Felt weird. 

- There are actually more than one way to represent percentages. We could use a `Double` with a value between 0 and 1 for instance. So here, we really mean to represent percentage in a "human friendly" way. My point is really that, most of the types we are going to represent using `MeasuredIn` will only ever have one unit (transaction fees are going to be in `lovelace/bytes` and that's not going to change, transaction max size will be in `bytes`, and same faith). The polymorphism gives us the ability to re-use the `MeasuredIn` type to represent various more concrete types. Likewise, we could just create a newtype for each one of them, but we'll end-up duplicating the serialization logic (and testing). So, yes, we could define a `Percentage` newtype and use helper to add units to the serialization and that will do the job. But then, we do not need the `MeasuredIn` abstraction and we could simply use newtype all the time. This is just a fight for consistency here.

- The "problem" with Percentage comes from the fact that we do not have any satisfactory primitive type to represent values between 0 and 100. I agree that the overlapping instance isn't satisfactory.

With both your feedbacks, this is the two suggestions / options I see:

- We define a `Percentage` or `Percentage100` primitive type that is basically an opaque type which can hold a non-floating value going from 0 to 100. And, we do use `MeasuredIn "percent" Percentage` as a type elsewhere to makes it obvious how this type is represented for us and, at the API-level. 

- Or, we add extra `n :: Nat` parameters to the `MeasuredIn` types to contains actual bounds instead of an underlying type. So we have: `newtype MeasuredIn (min :: Nat) (max :: Nat) (unit :: Symbol)`. With this, we can define bounds directly at the type-level and represent any arbitrary bounds we would need. It becomes quite unpractical to use and we probably end-up using type-alias like `type Percentage = MeasuredIn 0 100 "percent"`  ... :/

Reply:
I went for the first approach.

Reply:
> but we'll end-up duplicating the serialization logic (and testing)

`DerivingVia`?